ICSharpCode.Decompiler/IL/ControlFlow/SwitchDetection.cs

```diff
@@ -35,19 +36,121 @@ namespace ICSharpCode.Decompiler.IL.ControlFlow
 	/// </summary>
 	class SwitchDetection : IILTransform
 	{
-		SwitchAnalysis analysis = new SwitchAnalysis();
+		private readonly SwitchAnalysis analysis = new SwitchAnalysis();
+
+		private ILTransformContext context;
+		private BlockContainer currentContainer;
+		private ControlFlowGraph controlFlowGraph;
+		private LoopContext loopContext;
+
+		/// <summary>
+		/// When detecting a switch, it is important to distinguish Branch instructions which will
+		/// eventually decompile to continue; statements. 
+		/// 
+		/// A LoopContext is constructed for a node and its dominator tree, as for a Branch to be a continue;
+		/// statement, it must be contained within the target-loop
+		/// 
+		/// This class also supplies the depth of the loop targetted by a continue; statement relative to the
+		/// context node, to avoid (or eventually support) labelled continues to outer loops
+		/// </summary>
+		public class LoopContext
+		{
+			private readonly IDictionary<ControlFlowNode, int> continueDepth = new Dictionary<ControlFlowNode, int>();
+
+			public LoopContext(ControlFlowGraph cfg, ControlFlowNode contextNode)
+			{
```

Comment:
Can you comment some more on what this constructor is supposed to be doing?
Why is it analyzing the control flow reachable from the switch head, but stopping at loop heads?

Is it trying to follow the control flow up from the switch through the loop back-edge (a `continue;` statement) to the loop head, thus discovering the outer loop containing the switch?

If so, doesn't that fail for the following code?
```
while (true) {
  switch (i) {
     case 0:
        while (i < 10) i++;
        continue;
     default:
        return;
  }
}
```
The outer loop head is not reachable from the switch without going through an inner loop head, but your traversal stops at the inner loop head.

Reply:
Oh, `n.Dominates(contextNode)` is not the typical loop-head check. My example is no problem, since there's no loop head detected for the inner loop.
Still not quite sure which loop heads this constructor is supposed to find. All loop heads reachable from `contextNode` that also dominate the `contextNode`? It doesn't quite do that, though at least it should usually find the head of the loop that a `continue;` would target.

A different approach here would be to just iterate through the nodes dominating `contextNode`, and check if any of them is a loop head:
```
for (ControlFlowNode n = contextNode.ImmediateDominator; n != null; n = n.ImmediateDominator) {
   if (n.Predecessors.Any(p => n.Dominates(p))) {
      loopHeads.Add(n);
   }
}
```
This should give you the heads of all loops containing `contextNode`, already ordered from innermost to outermost.

Reply:
"heads of all loops containing `contextNode`" is the goal yes, so that continues (and their depth) can be identified.
I believe the example you've given will find loops which dominate the switch but don't contain it.

Reply:
The problem is that we're running at a point where loop bodies haven't been detected yet. "loops containing `contextNode`" is not quite defined yet.
Code dominated by a loop head could end up in the loop, even if there's no back-edge:
```
outer_loop_head:
while (...) {
    while (...) {
       if (...) {
         switch (...) { ... goto outer_loop_head; }
         return;
       }
   }
}
```
But such a block dominated by the loop without back-edge could also end up outside the loop -- ultimately it's up to loop detection's decision.
My code snipper approximates "heads of all loops containing contextNode" by computing "heads of all loops dominating contextNode" (thus giving an upper bound). Your code seems to give a lower bound instead, so really the question is: which type of error is better here, false positive or false negative?

Maybe we could change the pass ordering to detect loops before we start dealing with switches? I don't see a reason why `SwitchDetection` needs to run before `LoopDetection`; and loops->switches->ifs seems like it would cause less issues.

Reply:
The key is that if there's no back edge, then there's no continue statement, so you'll never encounter the false negative anyway.